### PR TITLE
Android custom icon and refactoring

### PR DESCRIPTION
**Pull request by bayrontq**
Hi guys!
I made some changes to the framework and would like you to revise before approving them:
- Added some methods for customizing notification icons on Android, since it lets you add the SmallIcon and LargeIcon.
- Refactored code that uses the Application class and the ApplicationContext method. To use them it received the Application as a parameter in ReactNativeUAPackage class. But this was not necessary because the ReactNative offers the getReactApplicationContext () method.

### CODE REVIEW

File: android/src/main/java/com/globo/reactnativeua/Preferences.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return instance;
+    }
+
+    private Preferences() {
+        preferences = UAirship.getApplicationContext().getSharedPreferences(
+                PREFS_NODE, Context.MODE_PRIVATE);
+    }
+
+    public int getAndroidSmallIconResourceId() {
+        return preferences.getInt(PREFS_ANDROID_SMALL_ICON, -1);
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/Preferences.java]: -1 is a valid resource id. Should default to 0.

File: android/src/main/java/com/globo/reactnativeua/Preferences.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public int getAndroidLargeIconResourceId() {
+        return preferences.getInt(PREFS_ANDROID_LARGE_ICON, -1);
+    }
+
+    public void setAndroidLargeIconResourceId(int iconResourceId) {
+        SharedPreferences.Editor editor = preferences.edit();
+        editor.putInt(PREFS_ANDROID_LARGE_ICON, iconResourceId);
+        editor.commit();
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/Preferences.java]: should use apply here. You can also chain these methods:

```
preferences.edit()
       .putInt(PREFS_ANDROID_LARGE_ICON, iconResourceId)
       .apply();
```

File: android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java [hunk 3]
```diff
@@ -156,4 +140,53 @@ private boolean shouldRequestPermissions(Activity activity) {
 
         return corseLocation == PackageManager.PERMISSION_DENIED && fineLocation == PackageManager.PERMISSION_DENIED;
     }
+
+    @ReactMethod
+    public void setAndroidSmallIcon(String iconName) {
+        int iconId = getImageResourceId(iconName);
+        if (iconId > 0) {
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java]: should do `if (iconId != 0)`

File: android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java [hunk 4]
```diff
@@ -156,4 +140,53 @@ private boolean shouldRequestPermissions(Activity activity) {
...
+            Preferences.getInstance().setAndroidLargeIconResourceId(iconId);
+            DefaultNotificationFactory defaultNotifFactory = getDefaultNotificationFactory();
+            defaultNotifFactory.setLargeIcon(iconId);
+            UAirship.shared().getPushManager().setNotificationFactory(defaultNotifFactory);
+        }
+    }
+
+    private DefaultNotificationFactory getDefaultNotificationFactory() {
+        final NotificationFactory notifFactory = UAirship.shared().getPushManager().getNotificationFactory();
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java:167]: Sorry about this.... In our 8.0 release we moved all the setters to the base class, we can clean this up later.

File: android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java [hunk 5]
```diff
@@ -156,4 +140,53 @@ private boolean shouldRequestPermissions(Activity activity) {
...
+        }
+        return new DefaultNotificationFactory(UAirship.getApplicationContext());
+    }
+
+    private int getImageResourceId(String imageName) {
+        if (imageName == null || imageName.length() <= 0) {
+            return -1;
+        }
+        int imageId = getImageResourceId(imageName, "drawable");
+        if (imageId <= 0) {
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/ReactNativeUA.java]: `if (imageId == 0)`

File: android/src/main/java/com/globo/reactnativeua/ReactNativeUAAutoPilot.java [hunk 6]
```diff
@@ -15,6 +17,25 @@ public void onAirshipReady(UAirship airship) {
 		UAirship.shared().getActionRegistry().unregisterAction(OpenExternalUrlAction.DEFAULT_REGISTRY_NAME);
 
         Log.w(TAG, "Airship ready!");
+
+        // Customize icons
+        DefaultNotificationFactory defaultNotifFactory = getDefaultNotificationFactory();
+        int smallIconResourceId = Preferences.getInstance().getAndroidSmallIconResourceId();
+        if (smallIconResourceId > 0) {
```
Comment (rlepinski) [inline android/src/main/java/com/globo/reactnativeua/ReactNativeUAAutoPilot.java]: `if (smallIconResourceId != 0)`

**Comment by rlepinski** [comment]
Android never specifies that negative resource values are invalid, only 0.

**Comment by bayrontq** [comment]
@rlepinski thanks for your help. I implemented the suggested changes.

**Pull request merged**